app/auth/oauth_client.py

```diff
@@ -0,0 +1,129 @@
+        """Overwrite parent method to work with provider properties
+        instead of an argument"""
+        return super().prepare_authorization_request(
+            self.provider_app.authorization_endpoint
+        )
+
+    def fetch_token(self, authorization_response, **kwargs):
+        """Convenience method for fetching tokens."""
+        oauth_session = OAuth2Session(client=self, redirect_uri=self.redirect_url)
+        oauth_session.fetch_token(
+            self.provider_app.token_endpoint,
+            authorization_response=authorization_response,
+            client_secret=self.provider_app.client_secret,
+            client_id=self.provider_app.client_id,
+            scopes=self.scopes,
+            include_client_id=True,
+            **kwargs
+        )
+        # TODO: Once Jupyterhub actually expires access tokens and implements
+        # TODO: usable refresh tokens, this subclass can be removed.
+        # Could use a factory method instead, but seems like an overkill...
+        if isinstance(self.provider_app, JupyterHubProviderApp):
+            self.refresh_token = None
+            self.expires_in = None
+        self._fix_expiration_time()
+
+    def refresh_access_token(self):
+        """Convenience method for refreshing tokens."""
+        self._expires_at = None
+        oauth_session = OAuth2Session(client=self)
+        oauth_session.refresh_token(
+            self.provider_app.token_endpoint,
+            client_id=self.provider_app.client_id,
+            client_secret=self.provider_app.client_secret,
+            include_client_id=True,
+        )
+        self._fix_expiration_time()
+
+    # Note: Pickling would be much simpler here, but we don't fully control
+    # what's going to be pickeled, so we choose the safer approach.
+    def to_json(self):
+        """Serialize a client into json."""
+        SERIALIZER_ATTRIBUTES = [
+            "token_type",
+            "access_token",
+            "refresh_token",
+            "token",
+            "scopes",
+            "state",
+            "code",
+            "redirect_url",
+            "max_lifetime",
+            "expires_in",
+            "_expires_at",
+        ]
+        client_dict = {key: vars(self)[key] for key in SERIALIZER_ATTRIBUTES}
+        client_dict["provider_app"] = self.provider_app.to_json()
+        return json.dumps(client_dict)
+
+    @classmethod
+    def from_json(cls, serialized_client):
+        """De-serialize a client from json."""
+        client_dict = json.loads(serialized_client)
+        client_dict["provider_app"] = OAuthProviderApp.from_json(
+            client_dict["provider_app"]
+        )
+        return cls(**client_dict)
+
+    def _fix_expiration_time(self):
+        """Cap a very long (or infinite) token lifetime. Note that we
+        do not modify the actual token (which is an attribute of the client
+        object) but instead let the client object expire."""
```

Comment:
Usually docs strings should follow a prescribed formatting from [pep8](https://www.python.org/dev/peps/pep-0257/#id17). Not sure if that's also a standard on the repo, but would be nice to have:

1) short sentence
2) blankline
3) elaborate description.

Reply:
Created an issue for this: https://github.com/SwissDataScienceCenter/renku-gateway/issues/233